### PR TITLE
[RPA-937] Implement multithreading

### PR DESCRIPTION
**Pull request by star-in-the-jar**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

---

I agree to the terms of the RunBotics Contributor License Agreement.

### CODE REVIEW

File: runbotics/runbotics-desktop/.env.development [hunk 1]
```diff
@@ -10,7 +10,8 @@ CFG_FIREFOX_BIN="C:\Program Files\Mozilla Firefox\firefox.exe"
 CFG_GECKO_DRIVER="./driver/windows/geckodriver-v0.29.1-win64"
 
 RUNBOTICS_LOGGER='winston'
-RUNBOTICS_EXTENSION_DIR='./extensions'
+RUNBOTICS_EXTENSION_DIR='C:\Users\A029616\code\runbotics-actions\'
```
Comment (star-in-the-jar) [inline runbotics/runbotics-desktop/.env.development]: i'll remove that line

**Review by star-in-the-jar — COMMENTED** [review]
self reviewed

File: runbotics/runbotics-desktop/.env.development (hunk 1 above)
Comment (preludium) [reply runbotics/runbotics-desktop/.env.development]: so remove it xd

File: runbotics/runbotics-scheduler/.env.development [hunk 2]
```diff
@@ -2,7 +2,7 @@ RUNBOTICS_ENTRYPOINT_URL='http://localhost:4000'
 RUNBOTICS_SCHEDULER_REDIS_HOST='localhost'
 RUNBOTICS_SCHEDULER_REDIS_PORT=6379
 
-RUNBOTICS_DB_HOST='localhost'
+RUNBOTICS_DB_HOST='127.0.0.1'
```
Comment (preludium) [inline runbotics/runbotics-scheduler/.env.development:5]: something went wrong here?

File: runbotics/runbotics-sdk/package.json [hunk 3]
```diff
@@ -10,12 +10,17 @@
...
+    "exports": {
+        ".": "./dist/index.js",
+        "./multithread/Worker": "./dist/worker.js"
+    },
     "scripts": {
         "postinstall": "rushx build",
-        "build": "microbundle --no-compress --format modern,cjs",
-        "start": "microbundle watch --no-compress --format modern,cjs"
+        "build": "microbundle src/**/*.ts --target node --format cjs",
+        "start": "microbundle watch --format modern,cjs"
```
Comment (preludium) [inline runbotics/runbotics-sdk/package.json]: adjust the `start` script accordingly to the `build` changes

File: runbotics/runbotics-sdk/src/handler/utils.ts [hunk 4]
```diff
@@ -4,25 +4,23 @@ import { StatefulActionHandler } from './StatefulActionHandler';
 export enum ActionHandler {
     Stateful = 'StatefulActionHandler',
     Stateless = 'StatelessActionHandler',
+    StatefulMultithreaded = 'StatefulMultithreaded',
+    StatelessMultithreaded = 'StatelessMultithreaded',
```
Comment (preludium) [inline runbotics/runbotics-sdk/src/handler/utils.ts]: adjust `Multithreaded` to `Multithread`

File: runbotics/runbotics-sdk/src/handler/utils.ts [hunk 5]
```diff
@@ -4,25 +4,23 @@ import { StatefulActionHandler } from './StatefulActionHandler';
...
-    && typeof handler === 'object'
-    && 'getType' in handler
-    && typeof handler.getType === 'function'
-    && handler.getType() === ActionHandler.Stateful
-    && 'tearDown' in handler
+): handler is StatefulActionHandler => typeof handler === 'object'
+&& 'getType' in handler
+&& typeof handler.getType === 'function'
+&& [ActionHandler.Stateful, ActionHandler.StatefulMultithreaded].includes(handler.getType())
+&& 'tearDown' in handler
```
Comment (preludium) [inline runbotics/runbotics-sdk/src/handler/utils.ts]: take a look at what is on the remote (null check is a bugfix) + create the same for 2 new types of handlers

File: runbotics/runbotics-sdk/src/multithread/Worker.ts [hunk 6]
```diff
@@ -0,0 +1,23 @@
+import { parentPort, workerData } from "worker_threads";
+import { WorkerMessage, WorkerMessageType } from "./Multithread.types";
+import { DesktopRunRequest } from "../handler/DesktopRunRequest";
+
+
+import(workerData.handlerPath)
+    .then(async (module) => {
+        const handler = module.default;
+        const handlerInstance = new handler();
```
Comment (preludium) [inline runbotics/runbotics-sdk/src/multithread/Worker.ts:9]: dont you need to use import/require here?

File: runbotics/runbotics-sdk/src/multithread/MultithreadRunner.ts [hunk 7]
```diff
@@ -0,0 +1,39 @@
+import { Worker } from 'worker_threads';
+import { DesktopRunRequest } from "../handler";
+import { MultithreadStateful } from "./MultithreadStateful";
+import { WorkerMessageType } from "./Multithread.types";
+
+export class MultithreadRunner extends MultithreadStateful {
+    private worker = null;
```
Comment (preludium) [inline runbotics/runbotics-sdk/src/multithread/MultithreadRunner.ts]: is there a type for worker instance?

File: runbotics/runbotics-sdk/src/multithread/MultithreadRunner.ts [hunk 8]
```diff
@@ -0,0 +1,39 @@
+import { Worker } from 'worker_threads';
+import { DesktopRunRequest } from "../handler";
+import { MultithreadStateful } from "./MultithreadStateful";
+import { WorkerMessageType } from "./Multithread.types";
+
+export class MultithreadRunner extends MultithreadStateful {
```
Comment (preludium) [inline runbotics/runbotics-sdk/src/multithread/MultithreadRunner.ts]: why does it extend multithreadStateful?

File: runbotics/runbotics-sdk/src/multithread/MultithreadStateless.ts [hunk 9]
```diff
@@ -0,0 +1,9 @@
+import { ActionHandler, DesktopRunRequest, DesktopRunResponse } from "../handler";
+
+export abstract class MultithreadStateless {
```
Comment (preludium) [inline runbotics/runbotics-sdk/src/multithread/MultithreadStateless.ts]: `MultithreadStatelessActionHandler`

File: runbotics/runbotics-sdk/src/multithread/MultithreadStateful.ts [hunk 10]
```diff
@@ -0,0 +1,11 @@
+import { ActionHandler, DesktopRunRequest, DesktopRunResponse } from "../handler";
+
+export abstract class MultithreadStateful {
```
Comment (preludium) [inline runbotics/runbotics-sdk/src/multithread/MultithreadStateful.ts]: `MultithreadStatefulActionHandler`

File: runbotics/runbotics-sdk/src/multithread/MultithreadStateful.ts [hunk 11]
```diff
@@ -0,0 +1,11 @@
...
+    abstract run(request: DesktopRunRequest): Promise<DesktopRunResponse>;
+
+    getType(): string {
+        return ActionHandler.StatefulMultithreaded;
```
Comment (preludium) [inline runbotics/runbotics-sdk/src/multithread/MultithreadStateful.ts]: keep one order `MultithreadStateful` vs `StatefulMultithread`

**Review by preludium — CHANGES_REQUESTED** [review]

File: runbotics/runbotics-sdk/src/multithread/MultithreadRunner.ts (hunk 7 above)
Comment (star-in-the-jar) [reply runbotics/runbotics-sdk/src/multithread/MultithreadRunner.ts]: i haven't found one

File: runbotics/runbotics-desktop/.env.development (hunk 1 above)
Comment (star-in-the-jar) [reply runbotics/runbotics-desktop/.env.development]: when I'm finished

File: runbotics/runbotics-sdk/src/handler/utils.ts [hunk 12]
```diff
@@ -4,25 +4,27 @@ import { StatefulActionHandler } from './StatefulActionHandler';
 export enum ActionHandler {
     Stateful = 'StatefulActionHandler',
     Stateless = 'StatelessActionHandler',
+    MultithreadStateful = 'MultithreadStateful',
+    MultithreadStateless = 'MultithreadStateless',
```
Comment (preludium) [inline runbotics/runbotics-sdk/src/handler/utils.ts]: missing `ActionHandler` postfix

File: runbotics/runbotics-sdk/src/handler/utils.ts [hunk 13]
```diff
@@ -4,25 +4,27 @@ import { StatefulActionHandler } from './StatefulActionHandler';
...
-    && handler.getType() === ActionHandler.Stateful
-    && 'tearDown' in handler
-    && typeof handler.tearDown === 'function';
+&& handler !== null
+&& typeof handler === 'object'
+&& 'getType' in handler
+&& typeof handler.getType === 'function'
+&& 'tearDown' in handler
+&& typeof handler.tearDown === 'function'
+    && [ActionHandler.Stateful, ActionHandler.MultithreadStateful].includes(handler.getType());
```
Comment (preludium) [inline runbotics/runbotics-sdk/src/handler/utils.ts]: keep this 1 tab indent + create separate functions for Multithread handlers, maybe we will need to distinct them in the future

File: runbotics/runbotics-sdk/src/multithread/MultithreadStatefulRunner.ts [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+    private worker = null;
+
+    constructor(handlerPath: string) {
+        super();
+        this.startWorker(handlerPath);
+    }
+
+    async tearDown(): Promise<void> {
+        return await this.worker.postMessage({ type: WorkerMessageType.EXIT });
+    }
```
Comment (preludium) [inline runbotics/runbotics-sdk/src/multithread/MultithreadStatefulRunner.ts]: no need to return anything here, just await for the end of promise

File: runbotics/runbotics-sdk/src/multithread/MultithreadStatefulRunner.ts [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+            this.onMessage(resolve, reject);
+        });
+    }
+
+    private async startWorker(handlerPath) {
+        this.worker = new Worker('./Worker.js', { workerData: { handlerPath } });
+    }
+
+    private async onMessage(resolve, reject, callback?: () => void) {
+        if (callback) callback();
```
Comment (preludium) [inline runbotics/runbotics-sdk/src/multithread/MultithreadStatefulRunner.ts]: what is this callback gonna do?

**Review by preludium — APPROVED** [review]

**Pull request merged**